wolfcrypt/src/logging.c

```diff
@@ -481,274 +430,517 @@ WOLFSSL_API int WOLFSSL_IS_DEBUG_ON(void)
-            }
-            #if defined(OPENSSL_EXTRA) && !defined(WOLFCRYPT_ONLY)
-            }
-            else {
-                (void)XSNPRINTF(buffer, sizeof(buffer),
-                    "wolfSSL error occurred, error = %d", error);
-            }
-            #endif
+#ifdef ERROR_QUEUE_PER_THREAD
+/* Keep the error queue in thread-local-storage. The only ways this
+ * API can deliver meaningful semantics in a multi-threaded setup.
+ */
+#ifndef ERROR_QUEUE_MAX
+/* Same as OpenSSL v1.1.x limit, note that this is per thread */
+#define ERROR_QUEUE_MAX 16
+#endif
 
-            wc_UnLockMutex(&debug_mutex);
-        }
-    #else
-        (void)XSNPRINTF(buffer, sizeof(buffer),
-                "wolfSSL error occurred, error = %d", error);
-    #endif
+struct wc_error_entry {
+    char   reason[WOLFSSL_MAX_ERROR_SZ];
+    char   file[WOLFSSL_MAX_ERROR_SZ];
+    int    line;
+    int    err;
+};
 
-    #ifdef DEBUG_WOLFSSL
-        if (loggingEnabled)
-            wolfssl_log(ERROR_LOG , buffer);
-    #endif
-    }
-}
+struct wc_error_queue {
+    struct wc_error_entry entries[ERROR_QUEUE_MAX];
+    size_t head_idx;
+    size_t count;
+};
 
-void WOLFSSL_ERROR_MSG(const char* msg)
-{
-#ifdef DEBUG_WOLFSSL
-    if (loggingEnabled)
-        wolfssl_log(ERROR_LOG , msg);
-#else
-    (void)msg;
-#endif
-}
+/* The complete queue in a thread local without allocations */
+static THREAD_LS_T struct wc_error_queue wc_errors;
 
-#endif  /* DEBUG_WOLFSSL || WOLFSSL_NGINX || WOLFSSL_HAPROXY */
+/* Using thread-local-storage, we do not need a mutex. */
+#define ERRQ_LOCK()            0
+#define ERRQ_UNLOCK()          (void)0
 
-#if defined(OPENSSL_EXTRA) || defined(DEBUG_WOLFSSL_VERBOSE)
 /* Internal function that is called by wolfCrypt_Init() */
 int wc_LoggingInit(void)
 {
-    if (wc_InitMutex(&debug_mutex) != 0) {
-        WOLFSSL_MSG("Bad Init Mutex");
-        return BAD_MUTEX_E;
-    }
-    wc_error_queue_count = 0;
-    wc_errors          = NULL;
-    wc_current_node    = NULL;
-    wc_last_node       = NULL;
-
     return 0;
 }
 
-
 /* internal function that is called by wolfCrypt_Cleanup */
 int wc_LoggingCleanup(void)
 {
     /* clear logging entries */
     wc_ClearErrorNodes();
-
-    /* free mutex */
-    if (wc_FreeMutex(&debug_mutex) != 0) {
-        WOLFSSL_MSG("Bad Mutex free");
-        return BAD_MUTEX_E;
-    }
     return 0;
 }
 
-
-/* peek at an error node
- *
- * idx : if -1 then the most recent node is looked at,
- *       otherwise search through queue for node at the given index starting
- *          from the absolute head wc_errors
- * file  : pointer to internal file string
- * reason : pointer to internal error reason
- * line  : line number that error happened at
- *
- * Returns a negative value in error case, on success returns the nodes error
- * value which is positive (absolute value)
- */
-int wc_PeekErrorNode(int idx, const char **file, const char **reason,
-        int *line)
+static int get_abs_idx(int relative_idx)
```

Comment:
these couple of internal functions could use a comment about what they return at the minimum

Reply:
will do